### PR TITLE
feat(core): support fuzzy highlighting

### LINKED ISSUE #4734: Error UI display in search results of quick search
### What happened?

When the search results contain text in 'Quote' format, there is a display error

image

### Distribution version

macOS ARM 64 (Apple Silicon)

### What browsers are you seeing the problem on if you're using web version?

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by JimmFly**
fix #4734 
add new fuzzy highlighting
image

### CODE REVIEW

File: packages/frontend/core/src/components/pure/cmdk/highlight.tsx [hunk 1]
```diff
@@ -24,33 +17,85 @@ type HighlightLabelProps = {
   highlight: string;
 };
 
+function useHighlight(text: string, query: string) {
```
Comment (pengx17) [inline packages/frontend/core/src/components/pure/cmdk/highlight.tsx:20]: i think it is better to have a `highlightTextFragments` function before wrapping it as a hook, and then write a unit test for `highlightTextFragments` alone.

File: packages/frontend/core/src/components/pure/cmdk/highlight.tsx [hunk 2]
```diff
@@ -24,33 +17,85 @@ type HighlightLabelProps = {
...
-    <div className={highlightContainer}>
-      {parts.map((part, i) => {
-        if (regex.test(part)) {
+    <div className={styles.highlightContainer}>
+      {highlights.map((part, i) => {
+        if (part.highlight) {
           return (
-            <span key={i} className={highlightKeyword}>
-              {part}
+            <span key={i} className={styles.highlightKeyword}>
```
Comment (joooye34) [inline packages/frontend/core/src/components/pure/cmdk/highlight.tsx:91]: use ternary operator to get class name, not the span element

**Review by joooye34 — DISMISSED** [review]

File: packages/frontend/core/src/hooks/__tests__/use-highlight.spec.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    const highlights = highlightTextFragments('This is a test', 'is');
+    expect(highlights).toStrictEqual([
+      { text: 'Th', highlight: false },
+      { text: 'is', highlight: true },
+      { text: ' is a test', highlight: false },
+    ]);
+  });
+
+  test('should correctly perform partial matching', () => {
+    const highlights = highlightTextFragments('Hello World', 'hw');
```
[the remark this answers is not in the document]
Comment (joooye34) [reply packages/frontend/core/src/hooks/__tests__/use-highlight.spec.ts:16]: Write a test to check multiple match, like `highlightTextFragments('My name is Nana,', 'na');`

File: packages/frontend/core/src/hooks/affine/use-highlight.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { useMemo } from 'react';
+
+export function highlightTextFragments(text: string, query: string) {
```
Comment (joooye34) [inline packages/frontend/core/src/hooks/affine/use-highlight.ts:3]: I am thinking maybe use iterator is more better.

File: packages/frontend/core/src/hooks/affine/use-highlight.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        text: text.substring(startIndex),
+        highlight: false,
+      });
+    }
+  }
+  return highlights;
+}
+
+export function useHighlight(text: string, query: string) {
+  return useMemo(() => highlightTextFragments(text, query), [text, query]);
```
Comment (joooye34) [inline packages/frontend/core/src/hooks/affine/use-highlight.ts:59]: if use iterator, this code cloud be `useMemo(() => Array.from(highlightTextFragments(text, query)), [text, query]);`

**Review by joooye34 — APPROVED** [review]

**Pull request merged**